### ISSUE TITLE
Socket timeout on Python 2.7 and 3.4

### ISSUE THREAD

**Issue by RunOrVeith** [sfalkner/pySMAC#18]
When running pySmac with python 2.7 on Ubuntu 14.04 with openJdk/Jre, a socket always times out after all rounds have been completed. It happens with the provided exaple files also.

Minimal example:

```
i = 1

def testfunction(x,y):
  global i
  print('Testing %i'%i)
  i = i+1
  return x+y

def testsmac():
  parameters = dict(x = ('real', [-1.,1.], 1),
                    y = ('real', [-1.,1.], 1),)
  opt = pysmac.SMAC_optimizer(debug=False)
  value, params = opt.minimize(testfunction, 5, parameters, num_procs = 1)
  print('Lowest: %f with %s'%(value, params))


if __name__ == '__main__':
  print('Starting optimization')
  start = time.time()
  testsmac()
  print('Needed %s seconds'%(time.time() - start))
```

to which the output is:

```

Starting optimization
Testing 1
Testing 1
Testing 1
Testing 1
Testing 1
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/pysmac/remote_smac.py", line 284, in remote_smac_function
    config_dict = smac.next_configuration()
  File "/usr/local/lib/python2.7/dist-packages/pysmac/remote_smac.py", line 193, in next_configuration
    self.__conn, addr = self.__sock.accept()
  File "/usr/lib/python2.7/socket.py", line 202, in accept
    sock, addr = self._sock.accept()
timeout: timed out
Lowest: -1.312829 with {'y': -0.7396858875859076, 'x': -0.5731433730945221}
Needed 6.80100393295 seconds

```

So in the end the result is still computed and returned, but the socket times out.
The timeout depends on the parameter "udp_timeout" defined in pysmac.remote_smac.
If that is increased, it waits for that time until the timout happens.
The output with debug=True enabled is attached.
I tried to find where that error originates and found that the timeoout occurs during the call
of 

  `pool.map(pysmac.remote_smac.remote_smac_function, argument_lists)`

in optimizer.py, but after all calls to pysmac.remote_smac.remote_smac_function have been executed. The subsequent call to pool.close() is executed after the timeout.

When trying to run the example above with python3.4,  I get the following error:

Edit: That error was a problem with the python3 virtualenv. The main issue (socket timeout) is also happening in python 3.4.

If more info is needed, I'm happy to provide it.
Thanks!
[testsmacDebug.txt](https://github.com/sfalkner/pySMAC/files/426757/testsmacDebug.txt)

**Comment by RunOrVeith** [comment]
Ok, I found a solution.

The problem is that the `next_configuration()` function of the `remote_smac` object is called one more time than the maximal allowed iterations (in the example above 6 instead of 5 times). Or better, it is called until`config_dict is None`, which only happens on purpose when an exception is caught once the connection to the socket has been established and then times out, but the timout on the connection establishment is not caught. 
On the last try, the socket accept times out, assumingly because the java SMAC has already begun to shut down since all iterations have finished(?). The default for python is then to return `None`, as no return statement is specified, thus the loop in remote_smac_function breaks.
I don't know why this apparently does not happen to everybody, especially because in the last step a noticeable amount of time (`udp_timeout` seconds) is wasted until the function returns `None`.

To get rid of the error, the max_evaluations parameter can be added into the argument_lists that is passed to the remote_smac_function, and then in line 283 of remote_smac.py we loop
over `while num_iterations < max_evaluations`, instead of `while True`.
This gets rid of the socket.timeout exception mentioned above.
Maybe (probababy) there is a better way to break the loop, but this works for now.
If this is good enough and this I can submit a pull request if you want, but I'm not sure if this is more of a hack or an actual solution.

Best regards

**Comment by jeslago** [comment]
I had the same issue. I tried your solution above, and it seems to work now.

**Comment by sfalkner** [comment]
I moved the socket query into the try/except block, which fixes that issue for me (see commit a8720a011cae970dcfb48cc10de92cf2e61f2e63)
